Need a monkey patch when generate a dashboard for SVM with hinge loss

`pycaret_classification.py", line 15, in save_dashboard
    dashboard = generate_classifier_explainer_dashboard(self.exp, self.best_model)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/*/Galaxy-pycaret/Galaxy-Pycaret/tools/dashboard.py", line 74, in generate_classifier_explainer_dashboard
    explainer = ClassifierExplainer(
                ^^^^^^^^^^^^^^^^^^^^
  File "/*/python3.11/site-packages/explainerdashboard/explainers.py", line 2462, in __init__
    assert hasattr(model, "predict_proba"), (
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: for ClassifierExplainer, model should be a scikit-learn compatible *classifier* model that has a predict_proba(...) method, so not a <class 'sklearn.linear_model._ridge.RidgeClassifier'>! If you are using e.g an SVM with hinge loss (which does not support predict_proba), you can try the following monkey patch:`

--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,6 @@
 .env
 env
+.venv
 *.log
 *.png
 *.txt

--- a/tools/base_model_trainer.py
+++ b/tools/base_model_trainer.py
@@ -265,8 +265,8 @@ class BaseModelTrainer:
                 Feature Importance</div>
             """
         if self.plots_explainer_html:
-            html_content += f"""
-                <div class="tab" onclick="openTab(event, 'explainer')">
+            html_content += """
+                "<div class="tab" onclick="openTab(event, 'explainer')">"
                 Explainer Plots</div>
             """
         html_content += f"""

--- a/tools/base_model_trainer.py
+++ b/tools/base_model_trainer.py
@@ -263,9 +263,13 @@ class BaseModelTrainer:
                 Best Model Plots</div>
                 <div class="tab" onclick="openTab(event, 'feature')">
                 Feature Importance</div>
+            """
+        if self.plots_explainer_html:
+            html_content += f"""
                 <div class="tab" onclick="openTab(event, 'explainer')">
-                Explainer
-                </div>
+                Explainer Plots</div>
+            """
+        html_content += f"""
             </div>
             <div id="summary" class="tab-content">
                 <h2>Setup Parameters</h2>
@@ -299,13 +303,19 @@ class BaseModelTrainer:
             <div id="feature" class="tab-content">
                 {feature_importance_html}
             </div>
+        """
+        if self.plots_explainer_html:
+            html_content += f"""
             <div id="explainer" class="tab-content">
                 {self.plots_explainer_html}
                 {tree_plots}
             </div>
-        {get_html_closing()}
-        """
-
+            {get_html_closing()}
+            """
+        else:
+            html_content += f"""
+            {get_html_closing()}
+            """
         with open(os.path.join(
                 self.output_dir, "comparison_result.html"), "w") as file:
             file.write(html_content)

--- a/tools/pycaret_classification.py
+++ b/tools/pycaret_classification.py
@@ -6,7 +6,7 @@ from dashboard import generate_classifier_explainer_dashboard
 
 from pycaret.classification import ClassificationExperiment
 
-from utils import add_hr_to_html, add_plot_to_html
+from utils import add_hr_to_html, add_plot_to_html, predict_proba
 
 LOG = logging.getLogger(__name__)
 
@@ -39,6 +39,16 @@ class ClassificationModelTrainer(BaseModelTrainer):
 
     def generate_plots(self):
         LOG.info("Generating and saving plots")
+
+        if not hasattr(self.best_model, "predict_proba"):
+            import types
+            self.best_model.predict_proba = types.MethodType(
+                predict_proba, self.best_model)
+            LOG.warning(
+                f"The model {type(self.best_model).__name__}\
+                    does not support `predict_proba`. \
+                    Applying monkey patch.")
+
         plots = ['confusion_matrix', 'auc', 'threshold', 'pr',
                  'error', 'class_report', 'learning', 'calibration',
                  'vc', 'dimension', 'manifold', 'rfe', 'feature',
@@ -74,9 +84,14 @@ class ClassificationModelTrainer(BaseModelTrainer):
         X_test = self.exp.X_test_transformed.copy()
         y_test = self.exp.y_test_transformed
 
-        explainer = ClassifierExplainer(self.best_model, X_test, y_test)
-        self.expaliner = explainer
-        plots_explainer_html = ""
+        try:
+            explainer = ClassifierExplainer(self.best_model, X_test, y_test)
+            self.expaliner = explainer
+            plots_explainer_html = ""
+        except Exception as e:
+            LOG.error(f"Error creating explainer: {e}")
+            self.plots_explainer_html = None
+            return
 
         try:
             fig_importance = explainer.plot_importances()

--- a/tools/pycaret_regression.py
+++ b/tools/pycaret_regression.py
@@ -59,9 +59,14 @@ class RegressionModelTrainer(BaseModelTrainer):
         X_test = self.exp.X_test_transformed.copy()
         y_test = self.exp.y_test_transformed
 
-        explainer = RegressionExplainer(self.best_model, X_test, y_test)
-        self.expaliner = explainer
-        plots_explainer_html = ""
+        try:
+            explainer = RegressionExplainer(self.best_model, X_test, y_test)
+            self.expaliner = explainer
+            plots_explainer_html = ""
+        except Exception as e:
+            LOG.error(f"Error creating explainer: {e}")
+            self.plots_explainer_html = None
+            return
 
         try:
             fig_importance = explainer.plot_importances()

--- a/tools/utils.py
+++ b/tools/utils.py
@@ -1,6 +1,8 @@
 import base64
 import logging
 
+import numpy as np
+
 logging.basicConfig(level=logging.DEBUG)
 LOG = logging.getLogger(__name__)
 
@@ -155,3 +157,8 @@ def encode_image_to_base64(image_path):
     """Convert an image file to a base64 encoded string."""
     with open(image_path, "rb") as img_file:
         return base64.b64encode(img_file.read()).decode("utf-8")
+
+
+def predict_proba(self, X):
+    pred = self.predict(X)
+    return np.array([1-pred, pred]).T

